File: 12nd/pro_3.java
import java.util.*;

class Solution {
    public int solution(String numbers) {
        int answer = 0;
        HashSet<Integer> set = new HashSet<>();
        permutation("",numbers,set);
        
        while(set.iterator().hasNext()){
            int a = set.iterator().next();
            set.remove(a);
            if(a==2) answer++;
            else if(isPrime(a)) answer++;
        }
        return answer;
    }
    
    public void permutation(String prefix, String str, HashSet<Integer> set){
        int n= str.length();
        if(!prefix.equals("")){
             set.add(Integer.valueOf(prefix));
        }
        for(int i=0;i<n; i++){
            permutation(prefix+str.charAt(i), str.substring(0,i)+str.substring(i+1,n),set);
        }
    }
    
    public boolean isPrime(int n){
        if(n==0||n==1)  return false;
        for(int i=2;i<=(int)Math.sqrt(n); i++){
            if(n%i==0)  return false;
        }
        return true;
    }
}

File: dp/pro_3.cpp
#include <string>
#include <vector>

using namespace std;

int f(int n){
    int dp[60001]={0,1,2,0,};
    for(int i=3; i<=n; i++){
        dp[i]=(dp[i-1]+dp[i-2])%1000000007;
    }
    return dp[n]%1000000007;
}

int solution(int n) {
    int answer = f(n);
    return answer;
}